File: game_environment2D_GA.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import abc
import tensorflow as tf
import numpy as np
from copy import deepcopy
from collections import deque

from tf_agents.environments import py_environment
from tf_agents.environments import tf_environment
from tf_agents.environments import tf_py_environment
from tf_agents.environments import utils
from tf_agents.specs import array_spec
from tf_agents.environments import wrappers
from tf_agents.environments import suite_gym
from tf_agents.trajectories import time_step as ts

tf.compat.v1.enable_v2_behavior()

class Organism:
    time_alive = 0
    x_pos = 0
    y_pos = 0
    energy = 0
    id = -1
    type = -1
    time_to_reproduce = 0
    coefficients = []
    polynomial_max_degree = 0
    def __multiply_params(self,parameters,times_used):
        result = 1.0
        for i in range(len(parameters)):
            result *= np.power(parameters[i],times_used[i])
        return result
        
    def __eval_function(self,parameters,times_used,position,max_degree,available_coefs):
        # Each recursive call will compute the sum of all parameter combinations
        # where each parameter is used max_degree times. Each call returns the sum of
        # all combinations starting with the given prefix - eg. if position = 3
        # the prefix is the first 3 parameters  
        if position >= len(parameters):
            return self.__multiply_params(parameters,times_used) * available_coefs.popleft()  
        
        sub_sum = 0.0
        for _ in range(max_degree+1):
            sub_sum += self.__eval_function(parameters,times_used,position+1,max_degree,available_coefs)
            times_used[position] += 1
        times_used[position] = 0
        return sub_sum
    
    def compute_function_recursive(self, parameters):
        times_used = [0]*len(parameters)
        available_coefs = deque(self.coefficients)
        return self.__eval_function(parameters,times_used,0,self.polynomial_max_degree,available_coefs)
        
    def compute_function_value(self,distance):
        distance_pow = 1
        result = 0.0
        for coef in self.coefficients:
            result += coef * distance_pow
            distance_pow *= distance
        return result

    def __init__(self, x_position, y_position, energy, type, time_to_reproduce, polynomial_degree = 0, coefficients = []):
        self.x_pos = x_position
        self.y_pos = y_position
        self.energy = energy
        self.type = type
        self.polynomial_max_degree = polynomial_degree
        self.coefficients = coefficients
        self.time_to_reproduce = time_to_reproduce
        self.time_alive = 0
        self.id = np.random.randint(-100000000,100000000)

    def to_list(self, include_coefs = False):
        if include_coefs:
            return [self.x_pos,self.y_pos,self.energy,self.time_to_reproduce,self.type,self.time_alive] + self.coefficients
        else:
            return [self.x_pos,self.y_pos,self.energy,self.time_to_reproduce,self.type,self.time_alive]

class GameEnv(py_environment.PyEnvironment):
  # Lists containing the currently alive organisms
  blue_organisms = []
  red_organisms = []

  # Lists containing the organisms that have died (used to pick the best organisms)
  dead_red_organisms = []
  dead_blue_organisms = []

  # Coefficients used when resetting (or starting) the game
  blue_starting_coefs = []
  red_starting_coefs = []
  polynomial_degree = 0

  green_organisms = []
  start_hp = 20 # HP / energy that all organisms have when the game starts
  board_length = 10 # Board will be of size board_length x board_length
  food_energy = 10 # Energy that each green organism will have
  max_moves = 200 # The maximum number of moves a game can last
  current_move_number = 0
  board_food_count = 10 # The number of green organisms always present on the board
  reproduction_cooldown = 3

  def __init__(self, blue_start_coefs, red_start_coefs, polynomial_degree, food_count = 10, board_size = 10):
    # Initialize specs
    self._action_spec = array_spec.BoundedArraySpec(
        shape=(), dtype=np.int32, minimum=0, maximum=3, name='action')
    self._observation_spec = array_spec.BoundedArraySpec(
        shape=(6*(self.board_food_count+len(blue_start_coefs)+len(red_start_coefs)),), dtype=np.int32, minimum=0, name='observation')

    self._episode_ended = False
    self.blue_organisms = []
    self.red_organisms = []
    self.dead_blue_organisms = []
    self.dead_red_organisms = []
    self.polynomial_degree = polynomial_degree
    self.board_food_count = food_count
    self.board_length = board_size

    # Initialize blue organisms
    for start_coefs in blue_start_coefs:
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)
        self.blue_organisms.append(Organism(x_pos,y_pos,self.start_hp,1,self.reproduction_cooldown,self.polynomial_degree,start_coefs))
    
    self.blue_starting_coefs = blue_start_coefs
    
    # Initialize red organisms
    for start_coefs in red_start_coefs:
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)
        self.red_organisms.append(Organism(x_pos,y_pos,self.start_hp,2,self.reproduction_cooldown,self.polynomial_degree,start_coefs))

    self.red_starting_coefs = red_start_coefs

    self.green_organisms = []
    # Generate green organisms
    for _ in range(self.board_food_count):
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)

        self.green_organisms.append(Organism(x_pos,y_pos,self.food_energy,0,self.reproduction_cooldown))

    self.current_move_number = 0

  def action_spec(self):
    return self._action_spec

  def observation_spec(self):
    return self._observation_spec

  def __organisms_to_array(self,organisms_list, starting_count):
    organism_array = []
    for i in range(min(starting_count,len(organisms_list))):
        organism_array += organisms_list[i].to_list()
    for _ in range(max(0,starting_count - len(organisms_list))):
        organism_array += [0]*6
    return organism_array

  def __get_current_game_state(self):
    state = []
    
    state += self.__organisms_to_array(self.blue_organisms, len(self.blue_starting_coefs))
    state += self.__organisms_to_array(self.red_organisms, len(self.red_starting_coefs))    

    for green_organism in self.green_organisms:
        state += green_organism.to_list()
    return state  

  def _reset(self):
    self._episode_ended = False

    self.blue_organisms = []
    self.dead_blue_organisms = []
    self.red_organisms = []
    self.dead_red_organisms = []
    
    for start_coefs in self.blue_starting_coefs:
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)
        self.blue_organisms.append(Organism(x_pos,y_pos,self.start_hp,1,self.reproduction_cooldown,self.polynomial_degree,start_coefs))
    
    for start_coefs in self.red_starting_coefs:
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)
        self.red_organisms.append(Organism(x_pos,y_pos,self.start_hp,2,self.reproduction_cooldown,self.polynomial_degree,start_coefs))

    self.green_organisms = []
    # Generate food
    for _ in range(self.board_food_count):
        x_pos = np.random.randint(0,self.board_length)
        y_pos = np.random.randint(0,self.board_length)

        self.green_organisms.append(Organism(x_pos,y_pos,self.food_energy,0,self.reproduction_cooldown))
    self.current_move_number = 0

    self._state = self.__get_current_game_state()
    return ts.restart(np.array(self._state, dtype=np.int32))

  def __reduce_organism_energy(self, organisms):
    for i in range(len(organisms)):
        organisms[i].energy -= 1
        organisms[i].time_to_reproduce = max(0,organisms[i].time_to_reproduce-1)
        organisms[i].time_alive += 1

  def __move_organism(self, organism, action):
        organism_copy = deepcopy(organism)
        # Process organism's action
        if action == 0:
            organism_copy.x_pos = (organism.x_pos-1+self.board_length)%self.board_length
        elif action == 1:
            organism_copy.x_pos = (organism.x_pos+1)%self.board_length
        elif action == 2:
            organism_copy.y_pos = (organism.y_pos+1)%self.board_length
        elif action == 3:
            organism_copy.y_pos = (organism.y_pos-1+self.board_length)%self.board_length
        else:
            raise ValueError('`action` should be in range [0,3].')
        return organism_copy

  def __get_distance(self,organism1,organism2):
    dx = organism1.x_pos - organism2.x_pos
    dy = organism1.y_pos - organism2.y_pos
    return abs(dx) + abs(dy)
  
  def __compute_organism_action(self,organism,other_organisms):
    if len(other_organisms) == 0:
        return 0
    # Find the organism that maximises the function
    max_result_organism = other_organisms[0]
    max_result = float('-inf')
    for other_organism in other_organisms:
        if organism.id == other_organism.id:
            continue
        distance = self.__get_distance(organism,other_organism)
        #function_value = organism.compute_function_value(distance)
        function_value = organism.compute_function_recursive([distance])
        if function_value > max_result:
            max_result = function_value
            max_result_organism = other_organism
    
    dx = organism.x_pos - max_result_organism.x_pos
    dy = organism.y_pos - max_result_organism.y_pos

    best_distance = dx*dx + dy*dy
    organism_action = 0
    possible_actions = [0,1,2,3]
    for action in possible_actions:
        new_organism = self.__move_organism(organism,action)
        distance = self.__get_distance(new_organism,max_result_organism)

        if distance <= best_distance and organism.type >= max_result_organism.type:
            organism_action = action
            best_distance = distance

        if distance >= best_distance and organism.type <= max_result_organism.type:
            organism_action = action
            best_distance = distance

    return organism_action
  
  # Computes and processes actions for organisms, computing function values for each organism in other_organisms
  def __process_actions_for_organisms(self,organisms, other_organisms):
    for i in range(len(organisms)):
        action = self.__compute_organism_action(organisms[i], other_organisms)
        organisms[i] = self.__move_organism(organisms[i],action)

  # Checks if any prey should be consumed, adds energy to the adequate predator organism
  # Sets energy = 0 for any prey that is consumed - dead prey must be removed afterwards
  def __consume_prey(self,predator_organisms, prey_organisms):
    for i in range(len(prey_organisms)):
        prey = prey_organisms[i]
        for j in range(len(predator_organisms)):
            predator = predator_organisms[j]
            if predator.x_pos == prey.x_pos and predator.y_pos == prey.y_pos:
                predator_organisms[j].energy += prey.energy
                prey_organisms[i].energy = 0
                break

  def __reproduce_organisms(self,organisms):
      for i in range(len(organisms)):
          for j in range(len(organisms)):
              if i != j:
                  if organisms[i].x_pos == organisms[j].x_pos and \
                  organisms[i].y_pos == organisms[j].y_pos and \
                  organisms[i].time_to_reproduce <= 0 and organisms[j].time_to_reproduce <= 0:
                      organisms[i].time_to_reproduce = self.reproduction_cooldown
                      organisms[j].time_to_reproduce = self.reproduction_cooldown
                      child_energy = organisms[i].energy/2 + organisms[j].energy/2
                      organisms[i].energy /= 2
                      organisms[j].energy /= 2
                      coefs = [0]*len(organisms[i].coefficients)
                      for coef_index in range(len(coefs)):
                          coefs[coef_index] = (organisms[i].coefficients[coef_index] + 
                                              organisms[j].coefficients[coef_index])/2.0
                      child = Organism(organisms[i].x_pos,organisms[i].y_pos,child_energy,organisms[i].type,
                                       self.reproduction_cooldown,organisms[i].polynomial_max_degree,coefs)
                      organisms.append(child)
  def _step(self, action):
    reward = 0
    if self._episode_ended:
      # The last action ended the episode. Ignore the current action and start
      # a new episode.
      return self.reset()

    self.__reduce_organism_energy(self.blue_organisms)
    self.__reduce_organism_energy(self.red_organisms)

    # Remove dead blue organisms
    self.dead_blue_organisms += [organism for organism in self.blue_organisms if organism.energy <= 0]
    self.blue_organisms = [organism for organism in self.blue_organisms if organism.energy > 0]
    # Remove dead red organisms
    self.dead_red_organisms += [organism for organism in self.red_organisms if organism.energy <= 0]
    self.red_organisms = [organism for organism in self.red_organisms if organism.energy > 0]

    # Make sure episodes don't go on forever.
    if self.current_move_number >= self.max_moves:
        self._episode_ended = True
        self.dead_blue_organisms += self.blue_organisms
        self.dead_red_organisms += self.red_organisms
    elif len(self.blue_organisms) == 0 or len(self.red_organisms) == 0:
        self._episode_ended = True
        if len(self.blue_organisms) > 0:
            self.dead_blue_organisms += self.blue_organisms
        if len(self.red_organisms) > 0:
            self.dead_red_organisms += self.red_organisms
        #reward = -20
    else:
        self.current_move_number += 1

    if not self._episode_ended:    
        all_organisms = self.green_organisms+self.red_organisms+self.blue_organisms
        self.__process_actions_for_organisms(self.blue_organisms,all_organisms)
        self.__process_actions_for_organisms(self.red_organisms,self.blue_organisms+self.red_organisms)
    
        reward = len(self.blue_organisms) + len(self.red_organisms)
        
        self.__consume_prey(self.blue_organisms,self.green_organisms)
        self.__consume_prey(self.red_organisms,self.blue_organisms)

        # Remove dead green and blue organisms. This must be done again 
        # for green and blue as additional organisms might have been consumed
        self.green_organisms = [organism for organism in self.green_organisms if organism.energy > 0]            
        self.dead_blue_organisms += [organism for organism in self.blue_organisms if organism.energy <= 0]
        self.blue_organisms = [organism for organism in self.blue_organisms if organism.energy > 0]            
    
        self.__reproduce_organisms(self.blue_organisms)
        self.__reproduce_organisms(self.red_organisms)
    
        # Generate new green organisms
        while len(self.green_organisms) != self.board_food_count:
            x_pos = np.random.randint(0,self.board_length)
            y_pos = np.random.randint(0,self.board_length)
            self.green_organisms.append(Organism(x_pos,y_pos,self.food_energy,0,self.reproduction_cooldown))

    self._state = self.__get_current_game_state()
    if self._episode_ended:
      return ts.termination(np.array(self._state, dtype=np.int32), reward)
    else:
      return ts.transition(
          np.array(self._state, dtype=np.int32), reward=reward, discount=0.9)

File: genetic_algorithm_cpp/game_environment2D_GA.cpp
#include "game_environment2D_GA.h"

bool TimeStep::is_last()
{
    return ts_type == "termination";
}

TimeStep::TimeStep(string type, double reward)
{
    ts_type = type;
    this->reward = reward;
}

double Organism::compute_function_recursive(vector<double>* parameters) const
{
    auto times_used = vector<int>(parameters->size(), -1);
    size_t coef_index = 0;
    int position = 0;
    double result = 0;
    vector<vector<double>> parameter_powers(parameters->size());
    for (size_t i = 0; i < parameter_powers.size(); i++)
    {
        parameter_powers[i] = vector<double>(polynomial_max_degree + 1);
        double current_pow = 1;
        double param = parameters->at(i);
        for (size_t j = 0; j < parameter_powers[i].size(); j++)
        {
            parameter_powers[i][j] = current_pow;
            current_pow *= param;
        }
    }

    while (position != -1)
    {
        if (position == times_used.size())
        {
            double params_multiplied = 1;
            for (size_t i = 0; i < parameters->size(); i++)
                params_multiplied *= parameter_powers[i][times_used[i]];

            //result += sigmoid(coefficients[coef_index++]*multiply_params(parameters,&times_used));
            result += sigmoid(coefficients[coef_index++] * params_multiplied);
            position--;
        }
        else
        {
            if (times_used[position] == polynomial_max_degree)
                times_used[position--] = -1;

            else
                times_used[position++]++;
        }
    }
    return result;
    /*int coef_pos = 0;
    auto times_used = vector<int>(parameters->size(), 0);
    auto res = eval_function(parameters, &times_used, 0, polynomial_max_degree, &coef_pos, coefficients);
    return res;*/
}

Organism::Organism(int x_position, int y_position, int energy, int type, int time_to_reproduce, int polynomial_degree, vector<double> coefficients)
{
    x_pos = x_position;
    y_pos = y_position;
    this->energy = energy;
    this->type = type;
    this->polynomial_max_degree = polynomial_degree;
    this->coefficients = coefficients;
    this->time_to_reproduce = time_to_reproduce;
    time_alive = 0;
    id = rand();
}

Organism::Organism()
{
    time_alive = 0;
}

vector<double> Organism::to_vector(bool include_coefs)
{
    vector<double> organism_vec = { (double)x_pos,(double)y_pos,energy,(double)time_to_reproduce,(double)type,(double)time_alive };
    if (include_coefs)
    {
        organism_vec.reserve(organism_vec.size() + coefficients.size());
        organism_vec.insert(organism_vec.end(), coefficients.begin(), coefficients.end());
    }
    return organism_vec;
}

double Organism::multiply_params(vector<double>* parameters, vector<int>* times_used) const
{
    double result = 1;
    for (int i = 0; i < parameters->size(); i++)
        result *= pow(parameters->at(i), times_used->at(i));

    return result;
}

double Organism::sigmoid(double value) const
{
    //return value;
    //return 1.0 / (1 + exp(-0.1*value));
    return value / (1 + abs(value));
}

double Organism::eval_function(vector<double>* parameters, vector<int>* times_used, int position, int max_degree, int* available_coef_pos, const vector<double>& coefs) const
{
    /* Each recursive call will compute the sum of all parameter combinations
    where each parameter is used max_degree times.Each call returns the sum of
    all combinations starting with the given prefix - eg. if position = 3
    the prefix is the first 3 parameters */
    if (position >= parameters->size())
    {
        double coef = coefs[*available_coef_pos];
        (*available_coef_pos)++;
        return sigmoid(multiply_params(parameters, times_used) * coef);
    }

    double sub_sum = 0;
    for (size_t i = 0; i < max_degree + 1; i++)
    {
        sub_sum += eval_function(parameters, times_used, position + 1, max_degree, available_coef_pos, coefs);
        times_used->at(position) += 1;
    }
    times_used->at(position) = 0;
    return sub_sum;
}

GameEnv::GameEnv(vector<vector<double>> blue_start_coefs, vector<vector<double>> red_start_coefs, int polynomial_degree, RandomUtil* rand_util, int food_count, int board_size)
{
    episode_ended = false;
    this->polynomial_degree = polynomial_degree;
    this->board_food_count = board_food_count;
    this->board_length = board_size;
    this->rand_util = rand_util;

    // Initialize blue organisms
    blue_organisms.reserve(blue_start_coefs.size() * 4);
    for (auto start_coefs : blue_start_coefs)
    {
        int x_pos = rand_util->rand_int(0, board_size - 1);
        int y_pos = rand_util->rand_int(0, board_size - 1);
        blue_organisms.push_back(Organism(x_pos, y_pos, start_hp, 1, reproduction_cooldown, polynomial_degree, start_coefs));
    }
    blue_starting_coefs = blue_start_coefs;

    // Initialize red organisms
    red_organisms.reserve(red_start_coefs.size() * 4);
    for (auto start_coefs : red_start_coefs)
    {
        int x_pos = rand_util->rand_int(0, board_size - 1);
        int y_pos = rand_util->rand_int(0, board_size - 1);
        red_organisms.push_back(Organism(x_pos, y_pos, start_hp, 2, reproduction_cooldown, polynomial_degree, start_coefs));
    }
    red_starting_coefs = red_start_coefs;

    // Generate green organisms
    green_organisms.reserve(board_food_count);
    for (size_t i = 0; i < board_food_count; i++)
    {
        int x_pos = rand_util->rand_int(0, board_size - 1);
        int y_pos = rand_util->rand_int(0, board_size - 1);
        green_organisms.push_back(Organism(x_pos, y_pos, food_energy, 0, reproduction_cooldown));
    }

    this->current_move_number = 0;
}

TimeStep GameEnv::reset()
{
    episode_ended = false;
    blue_organisms = vector<Organism>();
    red_organisms = vector<Organism>();
    green_organisms = vector<Organism>();
    dead_blue_organisms = list<Organism>();
    dead_red_organisms = list<Organism>();

    // Initialize blue organisms
    blue_organisms.reserve(blue_starting_coefs.size() * 4);
    for (auto start_coefs : blue_starting_coefs)
    {
        int x_pos = rand_util->rand_int(0, this->board_length - 1);
        int y_pos = rand_util->rand_int(0, this->board_length - 1);
        blue_organisms.push_back(Organism(x_pos, y_pos, start_hp, 1, reproduction_cooldown, polynomial_degree, start_coefs));
    }

    // Initialize red organisms
    red_organisms.reserve(red_starting_coefs.size() * 4);
    for (auto start_coefs : red_starting_coefs)
    {
        int x_pos = rand_util->rand_int(0, this->board_length - 1);
        int y_pos = rand_util->rand_int(0, this->board_length - 1);
        red_organisms.push_back(Organism(x_pos, y_pos, start_hp, 2, reproduction_cooldown, polynomial_degree, start_coefs));
    }

    // Generate green organisms
    green_organisms.reserve(board_food_count);
    for (size_t i = 0; i < board_food_count; i++)
    {
        int x_pos = rand_util->rand_int(0, this->board_length - 1);
        int y_pos = rand_util->rand_int(0, this->board_length - 1);
        green_organisms.push_back(Organism(x_pos, y_pos, food_energy, 0, reproduction_cooldown));
    }

    this->current_move_number = 0;
    return TimeStep("restart", 0);
}

void GameEnv::move_dead_organisms(vector<Organism>* organisms, list<Organism>* dead_organisms = nullptr)
{
    vector<Organism> alive_organisms; alive_organisms.reserve(organisms->size());
    for (int i = 0; i < organisms->size(); i++)
    {
        if (organisms->at(i).energy <= 0)
        {
            if (dead_organisms != nullptr)
                dead_organisms->push_back(organisms->at(i));
        }
        else
            alive_organisms.push_back(organisms->at(i));
    }
    *organisms = alive_organisms;
}

TimeStep GameEnv::step()
{
    double reward = 0;
    /* The last action ended the episode. Ignore the current action and start
    a new episode.*/
    if (episode_ended)
        return reset();

    reduce_organism_energy(&blue_organisms);
    reduce_organism_energy(&red_organisms);

    move_dead_organisms(&blue_organisms, &dead_blue_organisms);
    move_dead_organisms(&red_organisms, &dead_red_organisms);

    //Make sure episodes don't go on forever.
    if (current_move_number >= max_moves)
    {
        episode_ended = true;
        dead_blue_organisms.insert(dead_blue_organisms.end(), blue_organisms.begin(), blue_organisms.end());
        dead_red_organisms.insert(dead_red_organisms.end(), red_organisms.begin(), red_organisms.end());
    }
    else if (blue_organisms.size() == 0 || red_organisms.size() == 0)
    {
        episode_ended = true;
        if (blue_organisms.size() == 0)
        {
            /*for (size_t i = 0; i < red_organisms.size(); i++)
            {
                red_organisms[i].time_alive += red_organisms[i].energy;
            }*/
            dead_red_organisms.insert(dead_red_organisms.end(), red_organisms.begin(), red_organisms.end());
        }
        if (red_organisms.size() == 0)
        {
            /*for (size_t i = 0; i < blue_organisms.size(); i++)
            {
                blue_organisms[i].time_alive += blue_organisms[i].energy;
            }*/
            dead_blue_organisms.insert(dead_blue_organisms.end(), blue_organisms.begin(), blue_organisms.end());
        }
    }
    else
        current_move_number++;

    if (!episode_ended)
    {
        process_actions_for_organisms(&blue_organisms, { &green_organisms,&red_organisms,&blue_organisms });

        consume_prey(&blue_organisms, &green_organisms);
        consume_prey(&red_organisms, &blue_organisms);
        move_dead_organisms(&green_organisms);
        move_dead_organisms(&blue_organisms, &dead_blue_organisms);

        process_actions_for_organisms(&red_organisms, { &blue_organisms,&red_organisms });

        reward = blue_organisms.size() + red_organisms.size();

        consume_prey(&red_organisms, &blue_organisms);

        /*Remove dead green and blue organisms. This must be done again
        for green and blue as additional organisms might have been consumed*/
        move_dead_organisms(&blue_organisms, &dead_blue_organisms);

        reproduce_organisms(&blue_organisms);
        reproduce_organisms(&red_organisms);

        // Generate new green organisms
        green_organisms.reserve(board_food_count);
        while (green_organisms.size() != board_food_count)
        {
            int x_pos = rand_util->rand_int(0, this->board_length - 1);
            int y_pos = rand_util->rand_int(0, this->board_length - 1);
            green_organisms.push_back(Organism(x_pos, y_pos, food_energy, 0, reproduction_cooldown));
        }
    }
    if (episode_ended)
        return TimeStep("termination", reward);
    else
        return TimeStep("transition", reward);
}

void GameEnv::reduce_organism_energy(vector<Organism>* organisms)
{
    for (int i = 0; i < organisms->size(); i++)
    {
        organisms->at(i).energy -= 1;
        organisms->at(i).time_to_reproduce = max(0, organisms->at(i).time_to_reproduce - 1);
        organisms->at(i).time_alive += 1;
    }
}

void GameEnv::move_organism(Organism* organism, int action)
{
    // Process organism's action
    switch (action)
    {
    case 0:
        organism->x_pos = (organism->x_pos - 1 + board_length) % board_length;
        break;
    case 1:
        organism->x_pos = (organism->x_pos + 1) % board_length;
        break;
    case 2:
        organism->y_pos = (organism->y_pos + 1) % board_length;
        break;
    case 3:
        organism->y_pos = (organism->y_pos - 1 + board_length) % board_length;
        break;
    default:
        throw exception("action should be in range [0,3].");
    }
}

double GameEnv::get_distance(Organism* organism1, Organism* organism2)
{
    double dx = abs(organism1->x_pos - organism2->x_pos);
    double dy = abs(organism1->y_pos - organism2->y_pos);

    if (dx > board_length / 2)
        dx = board_length - dx;

    if (dy > board_length / 2)
        dy = board_length - dy;

    return abs(dx) + abs(dy);
}

pair<double, int> GameEnv::compute_organism_action(Organism* organism, vector<Organism>* other_organisms)
{
    double max_value = numeric_limits<double>::min();
    if (other_organisms->size() == 0)
        return { max_value, 0 };

    /*Organism* max_value_organism = &other_organisms->at(0);
    int best_action = 0;
    vector<int> possible_actions = { 0,1,2,3 };
    int start_x = organism->x_pos;
    int start_y = organism->y_pos;
    for (int i = 0; i < other_organisms->size(); i++)
    {
        auto other_organism = &other_organisms->at(i);
        if (other_organism->id == organism->id)
            continue;
        double distance = get_distance(organism, other_organism);
        vector<double> parameters = { distance };
        double function_value = organism->compute_function_recursive(&parameters);
        if (function_value > max_value)
        {
            max_value_organism = other_organism;
            max_value = function_value;
        }
    }
    bool is_predator = max_value_organism->type > organism->type;
    double best_distance = get_distance(organism, max_value_organism);
    for (int action: possible_actions)
    {
        move_organism(organism, action);
        double current_distance = get_distance(organism, max_value_organism);
        if (is_predator)
        {
            if (current_distance >= best_distance)
            {
                best_action = action;
                best_distance = current_distance;
            }
        }
        else
        {
            if (current_distance <= best_distance)
            {
                best_action = action;
                best_distance = current_distance;
            }
        }
        organism->x_pos = start_x;
        organism->y_pos = start_y;
    }*/

    int best_action = 0;
    vector<int> possible_actions = { 0,1,2,3 };
    int start_x = organism->x_pos;
    int start_y = organism->y_pos;

    for (int action : possible_actions)
    {
        move_organism(organism, action);
        double current_value = 0;
        for (int i = 0; i < other_organisms->size(); i++)
        {
            auto other_organism = &other_organisms->at(i);
            if (other_organism->id == organism->id)
                continue;

            double distance = get_distance(organism, other_organism);
            double is_predator_relative = 0;
            if (other_organism->type > organism->type)
            {
                is_predator_relative = 1;
            }
            else
                is_predator_relative = -1;
            vector<double> parameters = { distance, is_predator_relative, organism->energy };
            double function_value = organism->compute_function_recursive(&parameters);
            current_value += function_value;
        }
        if (current_value > max_value)
        {
            best_action = action;
            max_value = current_value;
        }
        organism->x_pos = start_x;
        organism->y_pos = start_y;
    }
    return { max_value,best_action };
}

void GameEnv::process_actions_for_organisms(vector<Organism>* organisms, vector<vector<Organism>*> other_organisms)
{
    for (int i = 0; i < organisms->size(); i++)
    {
        double max_score = numeric_limits<double>::min();
        int best_action = 0;
        for (auto other_organisms_vec : other_organisms)
        {
            auto best = compute_organism_action(&organisms->at(i), other_organisms_vec);
            if (best.first > max_score)
            {
                max_score = best.first;
                best_action = best.second;
            }
        }
        move_organism(&organisms->at(i), best_action);
    }
}

void GameEnv::consume_prey(vector<Organism>* predator_organisms, vector<Organism>* prey_organisms)
{
    for (int i = 0; i < prey_organisms->size(); i++)
    {
        auto prey = &prey_organisms->at(i);
        for (int j = 0; j < predator_organisms->size(); j++)
        {
            auto predator = &predator_organisms->at(j);
            if (predator->x_pos == prey->x_pos && predator->y_pos == prey->y_pos)
            {
                predator->energy += prey->energy;
                prey->energy = 0;
                break;
            }
        }
    }
}

void GameEnv::reproduce_organisms(vector<Organism>* organisms)
{
    vector<Organism> organisms_to_add;
    organisms_to_add.reserve(organisms->size() / 2);
    for (int i = 0; i < organisms->size(); i++)
    {
        for (int j = 0; j < organisms->size(); j++)
        {
            if (i != j)
            {
                if (organisms->at(i).x_pos == organisms->at(j).x_pos &&
                    organisms->at(i).y_pos == organisms->at(j).y_pos &&
                    organisms->at(i).time_to_reproduce <= 0 &&
                    organisms->at(j).time_to_reproduce <= 0)
                {
                    organisms->at(i).time_to_reproduce = reproduction_cooldown;
                    organisms->at(j).time_to_reproduce = reproduction_cooldown;
                    double child_energy = (organisms->at(i).energy + organisms->at(j).energy) / 2.0;
                    organisms->at(i).energy /= 2;
                    organisms->at(j).energy /= 2;
                    int coef_count = organisms->at(i).coefficients.size();
                    auto child_coefs = vector<double>(coef_count, 0);
                    for (int coef_index = 0; coef_index < coef_count; coef_index++)
                    {
                        child_coefs[coef_index] = (organisms->at(i).coefficients[coef_index] +
                            organisms->at(j).coefficients[coef_index]) / 2.0;
                    }
                    auto child = Organism(organisms->at(i).x_pos, organisms->at(i).y_pos, child_energy,
                        organisms->at(i).type, reproduction_cooldown, organisms->at(i).polynomial_max_degree,
                        child_coefs);
                    organisms_to_add.push_back(child);
                }
            }
        }
    }
    organisms->reserve(organisms->size() + organisms_to_add.size());
    organisms->insert(organisms->end(), organisms_to_add.begin(), organisms_to_add.end());
}
